### DISCUSSION TITLE
Multi-line comments accessibility for screen reader users

### DISCUSSION THREAD

**Discussion by Menelion** [community/community#7876] [Accessibility]
I brought it zillion times on various Microsoft accessibility channels, adding it here and hoping for good.

I use GitHub every day with a screen reader (my preferred setup is Windows + Chrome + JAWS for Windows, but I tested with different combinations). When reviewing a pull request, if you use the mouse, you can select several consecutive lines and add a comment or suggestion for all of them at once. Screen reader users and, in fact, all keyboard users (I assume) have no such possibility.

The solution I proposed is quite simple: There are line numbers. Make them interactable via keyboard (I'd say something like toggle buttons with some ARIA on them), so a press would either select or unselect a line. then, if a user clicks that plus button (which is correctly labeled as Add Line Comment, which was done upon my request back then), a comment will be spread to all the selected lines. If a non-contiguous line is selected, the selection would then be expanded so all the lines until that one would be selected. So, if I click number 5 and then number 10, all lines from 5 to 10 inclusive would be selected.

Thanks in advance.

**Comment by davorpa** [comment]
Great improve!!!

It also happens with touchable devices :wink:

**Reply by krzysztof293** [reply]
Ask??

**Comment by CyrilleB79** [comment]
Yes, it would be really useful, especially for projects such as [NVDA](https://github.com/nvaccess/nvda), developed for and partly by blind people people using keyboard only.

**Comment by Menelion** [comment]
@benjiallen This is the discussion I created about the multi-line comments accessibility. Would be nice to have this feature screen-reader accessible. Thanks!

**Reply by benjiallen** [reply]
Thanks for the feedback @Menelion I'll work with the PRs team to bring this to their attention and keep you in the loop when this lands on a roadmap.

CC: @willsmythe

**Reply by Menelion** [reply]
First, as this is still not implemented, @benjiallen is there any update on this? And second, could you probably move it to _Accessibility_ category? If I can do it on my own, please guide me how to do it. Thanks!

**Comment by akash07k** [comment]
Yes, it would be quite useful if gets implemented/fixed

**Comment by SpecOps12** [comment]
@benjiallen What is the state of this topic? Any news?

**Reply by benjiallen** [reply]
@SpecOps12 not much to share. The accessibility of commenting is a high priority for the team but there are a few things ahead of it right now. I'll post here once we get started as we'd love to get some user feedback on ideas we have.

**Reply by Menelion** [reply]
More than a year ago did I bring this issue for the first time. Actually close to one and a half years. It's bitter and sad to admit that such a simple accessibility improvement cannot make it to Github in such a huge amount of time.

**Reply by benjiallen** [reply]
@Menelion thanks for the feedback. We continue to work on it. Once we have ideas and solutions ready for more comment, we'll share.

**Reply by ShaneDittmar** [reply]
@benjiallen Has there been any update on making the multi-line comments feature accessible?

**Comment by Menelion** [comment]
And another half a year passed, even more. any updates on that?
P.S. The whole Accessibility category needs to be cleaned up, it's a nightmare actually. Lots of scam, off topic discussions and simply meaningless garbage.

**Reply by burmancomp** [reply]
There has been suggestion for one and half a year how this could be solved, and that suggestion has been voted up. However, nobody has had time to consider implementation for this in team.

I am waiting for constructive answer from the team when this is fixed. One and a half year is long time especially taking into account that solution has been suggested and voted up.

**Reply by XLTechie** [reply]
Any solution to this problem would be welcome!
As a blind developer, this has been a point of pain in the use of GitHub pull requests for years.
I am disappointed to find now that GitHub has been aware of it since at least 2021, but has not prioritized a solution.

Being unable to comment more than one line at a time in PR diffs, is both frustrating to me, and no doubt annoying to my sighted colleagues who must compensate for *"my" deficiencies* in only providing one-line suggestions that really need to be five or six line suggestions.

**Comment by cehfisher** [comment]
Hello @Menelion - thank you for taking the time to provide this feedback! Our Accessibility team is actively working on this issue. We will follow-up here with any additional questions or updates. Thanks again! (cc @queenofcorgis)

**Comment by cehfisher** [comment]
Hello @Menelion - thank you for your patience on this issue! With the recent updates to comments, the accessibility team is hopeful that this has now been fully resolved for you 🤞🏻🤞🏻🤞🏻

When you have a chance, could you re-test?

I’ve shared the team’s notes below in case they’re helpful:

> Validated the issue in current test environment and observed that issue is no longer reproducible as multiple lines can be selected in the diff via Shift + Up/Down key and the user can navigate on the 'Add Comment' button (+) to provide the comments on multiple lines at a time. Please refer to the attachment.
> 
> Test Environment
> OS: Windows 11 24H2 (OS Build: 26100.5074)
> Browser: Google Chrome (Version: 140.0.7339.81)

**Reply by Menelion** [reply]
@cehfisher Sorry, but everything is broken now. When I first opened GitHub, I didn't see any buttons at all, even with sighted help from my wife. 
<img width="1920" height="1008" alt="image" src="https://github.com/user-attachments/assets/b2dac4f8-b5aa-4288-9cd2-5d431e666487" />
Then I pressed "Try new experience". Here it's… still broken because although each line is reported as "clickable", I can't reliably add any comment, only in forms mode and with trial and error.  
Sorry to admit, but it's not usable at this time, and it's a great fortune I have no team work that requires commenting right now, otherwise I'd be in the deepest trouble.  
As you probably know, I'm absolutely opened for whatever, be it video calls, discussions, testing or anything else. I'm also ready to sign any NDA, should the need arise. Thanks.

**Comment by CyrilleB79** [comment]
@cehfisher, happy to see that this topic is moving forward.

I have tested with:
* Windows 11 24H2 (AMD64) build 26100.4946
* Firefox 142.0.1
* NVDA 2025.3beta3

I have no success when trying to get the "Comment" button, unless I move the mouse pointer to navigator object. And even when I do this, and select many lines with shift+up/downArrows and native selection enabled or not, only one line is being commented at a time.

Is this new multi-line comment capability documented somewhere for screen reader keyboard users so that I can follow it step-by-step?
Thanks

**Comment by cehfisher** [comment]
Hello @Menelion and @CyrilleB79 - thank you for the additional testing and feedback! I’ve asked the accessibility team and testers to re-open the internal issue and shared your comments and links with them to ensure it’s reviewed and resolved. We really appreciate your help in improving this.

**Comment by vojtapolasek** [comment]
Hello,
actually I must say that I think it works. But make sure that you enable the feature which is still in preview.
Go to your user menu -> preview feature -> click on "New files changed experience" and click "enable".
Now open a diff and move into the table. Switch to focus mode (I am using Orca on Linux, maybe it is called form mode in other screen readers?) and use arrows. You can move around the diff and also use shift+up / down to select lines. Access "add comment" button and other features by pressing tab / shift + tab. I will do more experiments and report here.

**Reply by CyrilleB79** [reply]
Hello
I am using the new preview interface for PRs now for some time. And I am satisfied with the new multi-line comment capability.

I also did not know that selecting multiple lines could be done with shift+up/down arrows. Thanks @vojtapolasek for the tip.

To summarize, the feature is now working as requested. Thanks for the good job!

Is there a documentation for this new GUI and its accessibility?

**Reply by CyrilleB79** [reply]
Answering myself:
The documentation is [here](https://accessibility.github.com/documentation/guide/pull-requests/), including the new experience, recommended for screen reader users.

Thanks for all. This topic can be closed from my point of view.

**Reply by cehfisher** [reply]
Thanks for the update @CyrilleB79!

What are your thoughts @Menelion? Did the recent updates help fix the issues you previously reported?

**Comment by cehfisher** [comment]
This issue has been re-validated and confirmed as resolved by the accessibility team. Below are some details for anyone who may still be experiencing related issues.

In the latest test environment, we verified that in the **new PR files changed UX**, users can now select multiple lines in the diff to add a comment by pressing **Shift + Up/Down arrow keys**, then selecting the **(+)** button to open the comment box.

In contrast, in the **old PR files changed UX**, users were unable to select multiple lines in the diff for commenting.

**Test environment:**

* **OS:** Windows 11 (Build 26200.6584)
* **Browser:** Google Chrome (Version 141.0.7390.66)
* **Assistive technologies:** NVDA (2025.3), JAWS (2025.2412.50)

**Reply by Menelion** [reply]
@cehfisher Sorry for being silent, I'm still thinking. So as for now, I can't review diff in virtual mode (JAWS) / browse mode (NVDA), only in forms/focus mode, right? I experience some problems with not hearing what's changed, added or removed. I need to redo a full assessment about it though.

**Reply by Menelion** [reply]
@cehfisher At least with JAWS, I have to jump back and forth between forms mode and virtual cursor mode (Focus/Browse in NVDA terminology). In Forms mode, especially in larger tables, it's kind of difficult to read the diff when there are blank lines in the Changed column. So yes, this kind of works, but it would be nice to discuss it further.

**Reply by XLTechie** [reply]
@Menelion Have you tried table navigation, instead of just `DownArrow`ing through the column? In NVDA, this is `Ctrl+Alt+Up/DownArrow`. That works better than just `Up/DownArrow` for navigating, at least in my (admittedly limited) testing.
I generally work with diffs in browse (virtual cursor) mode, as most of us presumably do, so this is a change of paradigm for me, but it seems to work so far, if a bit fiddly.

**Comment by CyrilleB79** [comment]
Also, it's quite unfortunate that the old control+G shortcut to provide a suggestion is not available anymore when writing a comment on one ore more line. Do you plan to restore it?

It frequently happens to me that I want to comment a line and then think to a suggestion. Then, I have to copy the whole comment that I have already written, discard the current comment then click to select the menu item to make a suggestion, and last, copy my comment above the suggestion pattern. It's quite unefficient and frustrating.

**Reply by XLTechie** [reply]
@CyrilleB79 While I agree that this feature should be returned (ironically, when mentioning accessibility improvements in the changelog, they don't happen to mention this removal...), there is an easier way.

If you have written a comment, return to browse mode, then move to and press the button for "Add a suggestion", it will be added at the end of your current comment field. Of course, it doesn't allow you to place it exactly where you want in the text, like `Ctrl+G` did, but at least you don't have to cut and paste what you've written already in order to get the suggestion in.

**Comment by XLTechie** [comment]
I would like to thank the technical team for working on this and providing a solution.
It took far more years than seems reasonable, but the fact that there is a definite way to do it now, is much appreciated!
